File: lectures/lecture06/code/lecture06.R
# =============================================================
# MANG3095 - Week 6: Fama and French (2015) five-factor model
# Full workflow: load Ken French factors, summary statistics,
# factor spanning regressions, GRS test, robustness checks.
# Data: ../../../data/csv/ff5_monthly.csv, ff5_annual.csv
# (monthly percent returns, July 1963 - August 2022, T = 710)
#
# Further resources used in the module material:
# - tidy-finance.org, chapter "Replicating Fama and French factors":
#   builds SMB/HML/RMW/CMA from CRSP + Compustat step by step in R.
# - github.com/ioannisrpt/FamaFrench2015FF5: replication package
#   for the FF (2015) paper (portfolio sorts and factor tables).
# =============================================================

df  <- read.csv("../../../data/csv/ff5_monthly.csv", check.names = FALSE)
dfa <- read.csv("../../../data/csv/ff5_annual.csv",  check.names = FALSE)
fac <- c("Mkt-RF", "SMB", "HML", "RMW", "CMA")
T   <- nrow(df)
cat(sprintf("Monthly factors: %s to %s, T = %d\n", df$Date[1], df$Date[T], T))

# ---- 2. Summary statistics (paper Table 4, Panel A) ----------
cat("\n=== Summary statistics, monthly % ===\n")
for (f in fac) {
  x <- df[[f]]
  cat(sprintf("%-7s mean %6.3f  sd %5.2f  t %5.2f  ann.Sharpe %5.2f\n",
              f, mean(x), sd(x), mean(x) / (sd(x) / sqrt(T)),
              mean(x) / sd(x) * sqrt(12)))
}
cat("\n=== Correlations ===\n")
print(round(cor(df[fac]), 2))

# ---- 3. Spanning regressions (paper Table 6) -----------------
# Regress each factor on the other four; a zero intercept means the
# factor is redundant for the mean-variance frontier of the others.
spanning <- function(target, data) {
  others <- setdiff(fac, target)
  y <- data[[target]]
  X <- as.matrix(data[others])
  lm(y ~ X)
}
cat("\n=== Spanning regressions ===\n")
for (f in fac) {
  m <- spanning(f, df)
  s <- summary(m)
  cat(sprintf("\n%s on the other four: alpha = %6.3f (t = %5.2f), R2 = %.2f\n",
              f, coef(s)[1, 1], coef(s)[1, 3], s$r.squared))
  print(round(coef(s), 3))
}
# Headline: HML's alpha is ~ -0.09 (|t| ~ 1.1): redundant once RMW
# and CMA are included. RMW and CMA keep alphas > 4 SE from zero.

# ---- 4. GRS test (Gibbons, Ross and Shanken 1989) -------------
grs_test <- function(assets, factors, data) {
  Fm <- as.matrix(data[factors]); Tn <- nrow(Fm); K <- ncol(Fm)
  N  <- length(assets)
  X  <- cbind(1, Fm)
  A  <- numeric(N); E <- matrix(0, Tn, N)
  for (i in seq_along(assets)) {
    m <- lm.fit(X, data[[assets[i]]])
    A[i] <- m$coefficients[1]; E[, i] <- m$residuals
  }
  Sigma <- crossprod(E) / (Tn - K - 1)
  mu    <- colMeans(Fm); Omega <- cov(Fm)
  grs <- (Tn - N - K) / N *
    drop(t(A) %*% solve(Sigma, A)) / (1 + drop(t(mu) %*% solve(Omega, mu)))
  p <- 1 - pf(grs, N, Tn - N - K)
  list(grs = grs, p = p, alphas = A)
}
cat("\n=== GRS test 1: RMW and CMA vs the FF3 factors ===\n")
g <- grs_test(c("RMW", "CMA"), c("Mkt-RF", "SMB", "HML"), df)
cat(sprintf("  alphas %.3f, %.3f; GRS = %.2f, p = %.2e -> reject FF3\n",
            g$alphas[1], g$alphas[2], g$grs, g$p))
cat("\n=== GRS test 2: HML vs Mkt, SMB, RMW, CMA ===\n")
g <- grs_test("HML", c("Mkt-RF", "SMB", "RMW", "CMA"), df)
cat(sprintf("  alpha %.3f; GRS = %.2f, p = %.3f -> cannot reject: redundant\n",
            g$alphas[1], g$grs, g$p))

# ---- 5. Robustness: pre/post 1991 split ----------------------
cat("\n=== Robustness: pre/post 1991 ===\n")
for (d in list(pre = df[df$Date < "1991-01", ], post = df[df$Date >= "1991-01", ])) {
  cat(sprintf("\n%s to %s (T = %d)\n", d$Date[1], d$Date[nrow(d)], nrow(d)))
  for (tgt in c("HML", "RMW", "CMA")) {
    s <- summary(spanning(tgt, d))
    cat(sprintf("  spanning %s: alpha = %6.3f (t = %5.2f), R2 = %.2f\n",
                tgt, coef(s)[1, 1], coef(s)[1, 3], s$r.squared))
  }
}

# ---- 6. Robustness: annual factors ---------------------------
cat("\n=== Robustness: annual factors ===\n")
for (f in fac) {
  x <- dfa[[f]]
  cat(sprintf("%-7s mean %6.2f  sd %5.2f  t %5.2f\n",
              f, mean(x), sd(x), mean(x) / (sd(x) / sqrt(nrow(dfa)))))
}
for (tgt in c("HML", "RMW", "CMA")) {
  s <- summary(spanning(tgt, dfa))
  cat(sprintf("annual spanning %s: alpha = %6.2f (t = %5.2f), R2 = %.2f\n",
              tgt, coef(s)[1, 1], coef(s)[1, 3], s$r.squared))
}

# What to interpret: all five premiums are positive over 1963-2022;
# the spanning alpha of HML is statistically zero (redundancy, FF 2015
# Section 7) while RMW and CMA are not; the GRS test rejects FF3
# against the new factors but not the four-factor model against HML;
# the redundancy result flips sign across the 1991 split, so treat it
# as sample specific, exactly as Fama and French caution.
# =============================================================

File: lectures/lecture06/code/lecture06.py
# =============================================================
# MANG3095 - Week 6: Fama and French (2015) five-factor model
# Full workflow: load Ken French factors, summary statistics,
# factor spanning regressions, GRS test, robustness checks.
# Data: ../../../data/csv/ff5_monthly.csv, ff5_annual.csv
# (monthly percent returns, July 1963 - August 2022, T = 710)
# =============================================================

import numpy as np
import pandas as pd
import statsmodels.api as sm
from scipy import stats

FACTORS = ["Mkt-RF", "SMB", "HML", "RMW", "CMA"]

# -------------------------------------------------------------
# 1. Load the factors
# -------------------------------------------------------------
df = pd.read_csv("../../../data/csv/ff5_monthly.csv")
dfa = pd.read_csv("../../../data/csv/ff5_annual.csv")
T = len(df)
print(f"Monthly factors: {df['Date'].iloc[0]} to {df['Date'].iloc[-1]}, T = {T}")
print(f"Annual factors : {dfa['Date'].iloc[0]} to {dfa['Date'].iloc[-1]}, T = {len(dfa)}")

# The factors are already excess returns in percent per month:
# Mkt-RF is the market return minus the T-bill rate, and SMB, HML,
# RMW, CMA are long-short (zero-investment) portfolio returns.
# Only RF (the riskfree rate itself) is not an excess return.

# -------------------------------------------------------------
# 2. Summary statistics (compare with paper Table 4, Panel A)
# -------------------------------------------------------------
print("\n=== Summary statistics, monthly % ===")
print(f"{'factor':8s} {'mean':>7s} {'sd':>6s} {'t(mean)':>8s} {'ann.Sharpe':>11s}")
for f in FACTORS:
    x = df[f]
    tstat = x.mean() / (x.std() / np.sqrt(T))
    sharpe = x.mean() / x.std() * np.sqrt(12)
    print(f"{f:8s} {x.mean():7.3f} {x.std():6.2f} {tstat:8.2f} {sharpe:11.2f}")

print("\n=== Factor correlations ===")
print(df[FACTORS].corr().round(2).to_string())

# -------------------------------------------------------------
# 3. Factor spanning regressions (paper Table 6):
#    regress each factor on the other four. If the intercept
#    (alpha) is zero, the factor adds nothing to the mean-
#    variance frontier spanned by the other four.
# -------------------------------------------------------------
def spanning(target, data):
    others = [f for f in FACTORS if f != target]
    m = sm.OLS(data[target], sm.add_constant(data[others])).fit()
    return m, others

print("\n=== Spanning regressions: each factor on the other four ===")
for f in FACTORS:
    m, others = spanning(f, df)
    print(f"\n{f} on {' + '.join(others)}:")
    print(f"  alpha = {m.params['const']:6.3f}  (t = {m.tvalues['const']:5.2f})   R2 = {m.rsquared:.2f}")
    for o in others:
        print(f"  {o:8s} {m.params[o]:6.3f}  (t = {m.tvalues[o]:6.2f})")

# Headline: the HML alpha is indistinguishable from zero once RMW and
# CMA are in the regression - HML is redundant (paper Section 7).

# -------------------------------------------------------------
# 4. GRS test (Gibbons, Ross and Shanken 1989)
#    Tests H0: all N intercepts are jointly zero when N test
#    assets are regressed on K factors.
# -------------------------------------------------------------
def grs_test(test_assets, factor_names, data):
    """GRS statistic and p-value. test_assets, factor_names: column lists."""
    F = data[factor_names].values                     # T x K
    Tn, K = F.shape
    N = len(test_assets)
    X = sm.add_constant(data[factor_names])
    alphas, resids = [], []
    for a in test_assets:
        m = sm.OLS(data[a], X).fit()
        alphas.append(m.params["const"])
        resids.append(m.resid.values)
    alphas = np.array(alphas)
    E = np.column_stack(resids)                       # T x N
    Sigma = E.T @ E / (Tn - K - 1)                    # residual covariance
    mu = F.mean(axis=0)
    Omega = np.cov(F.T, ddof=1)                       # factor covariance
    if Omega.ndim == 0:
        Omega = Omega.reshape(1, 1)
    quad_a = alphas @ np.linalg.solve(Sigma, alphas)
    quad_f = mu @ np.linalg.solve(Omega, mu)
    grs = (Tn - N - K) / N * quad_a / (1 + quad_f)
    p = 1 - stats.f.cdf(grs, N, Tn - N - K)
    return grs, p, alphas

print("\n=== GRS test 1: are RMW and CMA priced by the FF3 factors? ===")
g, p, al = grs_test(["RMW", "CMA"], ["Mkt-RF", "SMB", "HML"], df)
print(f"  FF3 alphas: RMW = {al[0]:.3f}, CMA = {al[1]:.3f}")
print(f"  GRS = {g:.2f}, p = {p:.2e}  -> reject: FF3 cannot price RMW and CMA")

print("\n=== GRS test 2: is HML priced by Mkt, SMB, RMW, CMA? ===")
g, p, al = grs_test(["HML"], ["Mkt-RF", "SMB", "RMW", "CMA"], df)
print(f"  four-factor alpha of HML = {al[0]:.3f}")
print(f"  GRS = {g:.2f}, p = {p:.3f}  -> cannot reject: HML is redundant")

# -------------------------------------------------------------
# 5. Robustness (a): subsample stability, split at 1991-01
# -------------------------------------------------------------
print("\n=== Robustness: pre/post 1991 split ===")
pre = df[df["Date"] < "1991-01"]
post = df[df["Date"] >= "1991-01"]
for label, d in [("pre-1991 ", pre), ("post-1991", post)]:
    print(f"\n{label} ({d['Date'].iloc[0]} to {d['Date'].iloc[-1]}, T = {len(d)})")
    for f in FACTORS:
        x = d[f]
        tstat = x.mean() / (x.std() / np.sqrt(len(d)))
        print(f"  {f:8s} mean {x.mean():6.3f} (t = {tstat:5.2f})")
    for tgt in ["HML", "RMW", "CMA"]:
        m, _ = spanning(tgt, d)
        print(f"  spanning {tgt}: alpha = {m.params['const']:6.3f} "
              f"(t = {m.tvalues['const']:5.2f}), R2 = {m.rsquared:.2f}")

# -------------------------------------------------------------
# 6. Robustness (b): annual factors vs monthly factors
# -------------------------------------------------------------
print("\n=== Robustness: annual factors (percent per year) ===")
Ta = len(dfa)
for f in FACTORS:
    x = dfa[f]
    tstat = x.mean() / (x.std() / np.sqrt(Ta))
    print(f"  {f:8s} mean {x.mean():6.2f}  sd {x.std():5.2f}  t = {tstat:5.2f}")
for tgt in ["HML", "RMW", "CMA"]:
    others = [f for f in FACTORS if f != tgt]
    m = sm.OLS(dfa[tgt], sm.add_constant(dfa[others])).fit()
    print(f"  annual spanning {tgt}: alpha = {m.params['const']:6.2f} "
          f"(t = {m.tvalues['const']:5.2f}), R2 = {m.rsquared:.2f}")

# -------------------------------------------------------------
# What to interpret:
# - Table 4 flavour: all five factor premiums are positive over the
#   full sample; CMA has the highest t-statistic, SMB the lowest.
# - Spanning: HML's four-factor alpha is around -0.09% per month with
#   |t| ~ 1.1: once RMW and CMA are included, HML adds nothing to the
#   description of average returns (FF 2015, Table 6 and Section 7).
#   RMW and CMA keep large, significant alphas: they are NOT redundant.
# - GRS: the joint test rejects FF3 against RMW+CMA overwhelmingly,
#   but cannot reject that HML's alpha is zero in the four-factor model.
# - Robustness: the redundancy result is sample specific. Pre-1991 the
#   HML spanning alpha is positive and significant (~0.21, t ~ 2.2);
#   post-1991 it is negative (~-0.31, t ~ -2.4). Annual-factor
#   regressions tell the same broad story as monthly ones.
# =============================================================
